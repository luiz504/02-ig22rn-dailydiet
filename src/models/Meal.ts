export type Meal = {
  id: string
  date: string
  name: string
  description: string
  inDiet: boolean
}
