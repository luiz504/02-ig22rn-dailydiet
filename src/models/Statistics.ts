export type Statistics = {
  inDietPercentage: number
  bestInDietSequence: number
  inDietEntries: number
  outDietEntries: number
  totalEntries: number
}
