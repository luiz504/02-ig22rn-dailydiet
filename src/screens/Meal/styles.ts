import styled from 'styled-components/native'

export const Row = styled.View`
  gap: 8px;
  margin-bottom: 24px;
`
export const DietCard = styled.View`
  padding: 8px 16px;
  border-radius: 999px;

  gap: 8px;
  background-color: ${({ theme }) => theme.colors['gray-200']};

  flex-direction: row;
  align-self: flex-start;
  align-items: center;
`
export const Dot = styled.View`
  width: 8px;
  height: 8px;
  border-radius: 999px;
`

export const Col = styled.View`
  margin-top: auto;
  gap: 8px;
`
