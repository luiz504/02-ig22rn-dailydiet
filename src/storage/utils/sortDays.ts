export const sortDays = (array: string[]) => {
  return array.sort((a, b) => {
    return new Date(b).getTime() - new Date(a).getTime()
  })
}
