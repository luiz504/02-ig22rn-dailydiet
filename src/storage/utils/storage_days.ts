import AsyncStorage from '@react-native-async-storage/async-storage'
import { DAY_COLLECTION } from '../commons'

export async function getStoredDays() {
  return AsyncStorage.getItem(DAY_COLLECTION).then((storedData) =>
    storedData ? (JSON.parse(storedData) as string[]) : null,
  )
}

export async function setStoredDays(dayCollection: string[]) {
  return AsyncStorage.setItem(DAY_COLLECTION, JSON.stringify(dayCollection))
}
