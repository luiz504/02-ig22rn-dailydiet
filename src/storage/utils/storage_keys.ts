import { format, parse } from 'date-fns'

const FORMAT = 'yyyy/MM/dd'
export function storageDateKeyFormat(date: Date): string {
  return format(date, FORMAT)
}

export function storageKeyToDateParse(key: string) {
  return parse(key, FORMAT, new Date())
}
