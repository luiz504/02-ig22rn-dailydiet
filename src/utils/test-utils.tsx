import { FC, ReactNode } from 'react'
import { ThemeProvider } from 'styled-components/native'
import { render } from '@testing-library/react-native'

import { theme } from '~/styles'
import { NavigationContainer } from '@react-navigation/native'

const AllTheProviders: FC<{ children: ReactNode }> = ({ children }) => {
  return (
    <ThemeProvider theme={theme}>
      <NavigationContainer>{children}</NavigationContainer>
    </ThemeProvider>
  )
}

const ThemedWrapper: FC<{ children: ReactNode }> = ({ children }) => {
  return <ThemeProvider theme={theme}>{children}</ThemeProvider>
}

const customRender: typeof render = (ui, options) =>
  render(ui, { wrapper: AllTheProviders, ...options })

const renderThemed: typeof render = (ui, options) =>
  render(ui, { wrapper: ThemedWrapper, ...options })

const splitBorderColor = (borderColor: string) => ({
  borderTopColor: borderColor,
  borderRightColor: borderColor,
  borderBottomColor: borderColor,
  borderLeftColor: borderColor,
})

export * from '@testing-library/react-native'

export { customRender as render, renderThemed, splitBorderColor }
