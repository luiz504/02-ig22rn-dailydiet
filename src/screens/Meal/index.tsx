import { useNavigation, useRoute } from '@react-navigation/native'
import { FC, useState } from 'react'
import { useTheme } from 'styled-components/native'

import { formatDateAndTime } from '~/utils/dateTimeFormatters'

import { ContentSection } from '~/components/ContentSection'
import { HeaderShort } from '~/components/HeaderShort'
import { Text } from '~/components/Text'
import { Theme } from '~/components/Theme'
import { Button } from '~/components/Button'
import { AlertDelete } from './components/AlertDelete'

import { Row, DietCard, Dot, Col } from './styles'

import { Meal } from '~/models/Meal'
import { Alert } from 'react-native'

import { deleteMeal } from '~/storage/meals/deleteMeal'

type RouteParams = {
  meal: Meal
  groupName: string
}

export const MealScreen: FC = () => {
  const navigator = useNavigation()
  const theme = useTheme()
  const { meal, groupName } = useRoute().params as RouteParams

  const themeVariant = meal.inDiet ? 'green' : 'red'

  const dietCardLabel = meal.inDiet ? 'within the diet' : 'outside the diet'
  const dotColor = meal.inDiet
    ? theme.colors['green-900']
    : theme.colors['red-900']

  const dateAndTimeString = formatDateAndTime(meal.date)

  const handleEditMeal = () => {
    navigator.navigate('edit-meal', { meal, groupName })
  }
  const [isDeleting, setIsDeleting] = useState(false)
  const handleDeleteMeal = async () => {
    try {
      setIsDeleting(true)
      await deleteMeal({ groupName, mealId: meal.id })

      navigator.navigate('home')
    } catch (err) {
      setIsDeleting(false)

      Alert.alert(
        'Delete Meal Error',
        'Something went wrong with the deletion request.',
      )
    }
  }

  const [showDeleteAlert, setShowDeleteAlert] = useState(false)
  const handleRequestDeleteMeal = (type: 'request' | 'abort') => {
    setShowDeleteAlert(type === 'request')
  }

  return (
    <>
      <Theme variant={themeVariant} testID="container">
        <HeaderShort
          title="Meal"
          onReturnRequest={() => navigator.navigate('home')}
          disabledReturn={isDeleting}
        />

        <ContentSection>
          <Row>
            <Text
              weight="bold"
              style={{ fontSize: 20, lineHeight: 26 }}
              testID="text-name"
            >
              {meal.name}
            </Text>

            <Text testID="text-description">{meal.description}</Text>
          </Row>

          <Row>
            <Text weight="bold" size="sm">
              Date and time
            </Text>

            <Text testID="text-date-time">{dateAndTimeString}</Text>
          </Row>

          <DietCard testID="diet-status-card">
            <Dot
              style={{ backgroundColor: dotColor }}
              testID="diet-status-dot"
            />
            <Text size="sm" testID="diet-status-label">
              {dietCardLabel}
            </Text>
          </DietCard>

          <Col>
            <Button
              icon="pen"
              label="Edit Meal"
              onPress={handleEditMeal}
              disabled={isDeleting}
              testID="btn-edit"
            />

            <Button
              icon="trash"
              variant="outline"
              label="Delete Meal"
              onPress={() => handleRequestDeleteMeal('request')}
              disabled={isDeleting}
              testID="btn-delete"
            />
          </Col>
        </ContentSection>
      </Theme>

      <AlertDelete
        show={showDeleteAlert}
        onDismiss={() => handleRequestDeleteMeal('abort')}
        onConfirm={handleDeleteMeal}
      />
    </>
  )
}
