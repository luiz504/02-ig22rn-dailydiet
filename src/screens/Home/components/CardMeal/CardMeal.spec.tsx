import { render, screen } from '~/utils/test-utils'
import { CardMeal } from '.'
import { Meal } from '~/models/Meal'

describe('CardMeal Component', () => {
  const mockMeals: Meal[] = [
    {
      id: 'm1',
      name: 'fake Meal',
      description: 'fake Description',
      date: new Date().toISOString(),
      inDiet: true,
    },
    {
      id: 'm2',
      name: 'fake Meal 2',
      description: 'fake Description 2',
      date: new Date().toISOString(),
      inDiet: false,
    },
  ]

  const containerID = 'card'
  it('should render correctly with meal props property "inDiet" as true', () => {
    const currentMeal = mockMeals[0]
    render(<CardMeal meal={currentMeal} testID={containerID} />)

    const containerBtnElement = screen.getByTestId(containerID)
    const hourTextElement = screen.getByTestId('hour')
    const hrElement = screen.getByTestId('hr')
    const nameTextElement = screen.getByText(currentMeal.name)
    const statusElement = screen.getByTestId('status')

    expect(containerBtnElement.props.style).toMatchSnapshot('container-style')
    expect(hourTextElement.props.style).toMatchSnapshot('hour-text-style')
    expect(hrElement.props.style).toMatchSnapshot('hr-style')
    expect(nameTextElement.props.style).toMatchSnapshot('name-text-style')
    expect(nameTextElement.props.numberOfLines).toBe(1)
    expect(statusElement.props.style).toMatchSnapshot('status-style')
  })

  it('should render correctly with meal props property "inDiet" as false', () => {
    const currentMeal = mockMeals[1]
    render(<CardMeal meal={currentMeal} testID={containerID} />)

    const statusElement = screen.getByTestId('status')

    expect(statusElement.props.style).toMatchSnapshot('status-style')
  })

  it('should render correctly the HH:mm time format', () => {
    const currentMeal = mockMeals[0]
    render(<CardMeal meal={currentMeal} testID={containerID} />)
    const hourTextElement = screen.getByTestId('hour')

    const regex = /^(?:[01]\d|2[0-3]):[0-5]\d$/

    expect(hourTextElement).toHaveTextContent(regex)
  })
})
