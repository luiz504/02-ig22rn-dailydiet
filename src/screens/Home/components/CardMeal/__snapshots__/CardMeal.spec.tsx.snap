// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardMeal Component should render correctly with meal props property "inDiet" as false: status-style 1`] = `
[
  {
    "borderBottomLeftRadius": 9999,
    "borderBottomRightRadius": 9999,
    "borderTopLeftRadius": 9999,
    "borderTopRightRadius": 9999,
    "height": 14,
    "width": 14,
  },
  {
    "backgroundColor": "#F3BABD",
  },
]
`;

exports[`CardMeal Component should render correctly with meal props property "inDiet" as true: container-style 1`] = `
{
  "alignItems": "center",
  "borderBottomColor": "#DDDEDF",
  "borderBottomLeftRadius": 6,
  "borderBottomRightRadius": 6,
  "borderBottomWidth": 1.5,
  "borderLeftColor": "#DDDEDF",
  "borderLeftWidth": 1.5,
  "borderRightColor": "#DDDEDF",
  "borderRightWidth": 1.5,
  "borderTopColor": "#DDDEDF",
  "borderTopLeftRadius": 6,
  "borderTopRightRadius": 6,
  "borderTopWidth": 1.5,
  "flexDirection": "row",
  "gap": 12,
  "opacity": 1,
  "paddingBottom": 14,
  "paddingLeft": 12,
  "paddingRight": 16,
  "paddingTop": 14,
}
`;

exports[`CardMeal Component should render correctly with meal props property "inDiet" as true: hour-text-style 1`] = `
{
  "color": "#1B1D1E",
  "fontFamily": "Nunito-Bold",
  "fontSize": 12,
  "lineHeight": 15.6,
}
`;

exports[`CardMeal Component should render correctly with meal props property "inDiet" as true: hr-style 1`] = `
{
  "backgroundColor": "#B9BBBC",
  "height": "100%",
  "width": 1,
}
`;

exports[`CardMeal Component should render correctly with meal props property "inDiet" as true: name-text-style 1`] = `
[
  {
    "color": "#333638",
    "fontFamily": "Nunito-Regular",
    "fontSize": 16,
    "lineHeight": 20.8,
  },
  {
    "flex": 1,
  },
]
`;

exports[`CardMeal Component should render correctly with meal props property "inDiet" as true: status-style 1`] = `
[
  {
    "borderBottomLeftRadius": 9999,
    "borderBottomRightRadius": 9999,
    "borderTopLeftRadius": 9999,
    "borderTopRightRadius": 9999,
    "height": 14,
    "width": 14,
  },
  {
    "backgroundColor": "#CBE4B4",
  },
]
`;
